### ISSUE TITLE
[Test] 30-Round Adversarial Audit Report

### ISSUE THREAD

**Issue by SRjoeee** [Findit-AI/soundevents#8]
# Audit Report: soundevents crate

**Date:** 2026-05-09
**Crate version:** 0.3.1 (soundevents), 0.2.0 (soundevents-dataset)
**Rust edition:** 2024 (soundevents), 2021 (soundevents-dataset)
**MSRV:** 1.85 (soundevents), 1.59 (soundevents-dataset)
**Total source lines:** ~18K hand-written + ~30K generated

---

## Executive Summary

The soundevents crate is a production-oriented Rust inference wrapper for CED
(AudioSet) sound event classifiers. It provides ONNX Runtime inference with
support for single-clip, batch, and chunked classification over 527 AudioSet
classes. The codebase is well-structured, thoroughly documented, and has strong
test coverage (41 tests). The overall quality is **high** with a few minor
observations noted below.

---

### Architecture

- Uses `ort` crate v2.0.0-rc.12 (ONNX Runtime bindings)
- Single `Session` per `Classifier` instance
- Input tensor: `[batch_size, num_samples]` — raw 16kHz mono f32 waveform
- Output tensor: `[batch_size, 527]` — raw logits (sigmoid applied post-inference)
- `SmolStr` used for I/O tensor name storage (avoids heap allocation for short names)

### Findings

**Good:**
- Proper error propagation from ORT via `#[from] ort::Error`
- Session builder allows configuring `GraphOptimizationLevel`
- Supports both file-based and in-memory model loading
- `bundled-tiny` feature embeds a 6.4MB ONNX model for zero-config usage
- Output validation is strict: checks shape rank, batch size, class count

**Minor concerns:**

1. **`with_optimization_level` error mapping (line 773-774):**
   ```rust
   .with_optimization_level(optimization_level)
   .map_err(ort::Error::from)?
   ```
   This suggests the ORT builder returns a non-standard error type. The
   conversion is correct but fragile if the ORT API changes. A comment
   explaining why this is needed would be helpful.

2. **Only first I/O tensor used (lines 781-794):** The code takes
   `.inputs().first()` and `.outputs().first()`. If a model exposes multiple
   inputs or outputs, only the first is used. This is correct for CED models
   but not validated at model-load time. A debug assertion or warning for
   multi-input models would be prudent.

3. **No warmup run:** The first inference call after loading will be slower
   due to ORT memory allocation. A `warmup()` method or documentation note
   about this would be useful for latency-sensitive applications.

---

### Design

- **No preprocessing is performed.** Raw 16kHz mono f32 samples are passed
  directly to the ONNX model.
- This is by design: CED models internally compute mel-spectrogram features.
- The crate expects callers to handle resampling and channel mixing.

### Findings

**Good:**
- Clean API contract: `SAMPLE_RATE_HZ = 16_000` constant documents the expectation
- `EmptyInput` error for zero-length audio
- No hidden allocations for preprocessing

**Observations:**

4. **No sample rate validation:** The crate trusts that input is 16kHz. Passing
   44.1kHz audio will produce incorrect results silently. A documentation note
   in the top-level doc comment would be valuable. Consider adding a
   `debug_assert!` or optional validation method.

5. **No amplitude range check:** The crate accepts any f32 values. CED models
   typically expect normalized audio in [-1, 1]. Out-of-range values will
   produce unexpected results silently.

---

### Design

- `chunk_slices()`: divides audio into fixed-size windows with configurable hop
- Default: 10-second windows (160,000 samples), non-overlapping
- Tail chunks are shorter (no zero-padding) — intentional design choice
- `chunk_batches()` groups consecutive equal-length chunks into batches
  (prevents mixing tail chunks with full-size windows)

### Findings

**Good:**
- Overlapping windows supported via `hop_samples < window_samples`
- Batching correctly handles the tail chunk boundary
- Mean aggregation accumulates during iteration, divides at end (numerically
  better than per-chunk division)
- Max aggregation uses per-class `f32::max`
- Zero window/hop/batch validated via `validate_chunking()`

**Observations:**

6. **Tail chunk produces different model behavior:** A 2-second tail chunk fed
   to a model trained on 10-second windows may produce different per-second
   detection characteristics. This is documented behavior but worth noting in
   the doc comments for `classify_chunked`.

7. **Mean aggregation denominator counts all chunks equally:** If
   `hop < window` (overlapping), shorter tail chunks get the same weight as
   full windows in the mean. This is the standard approach but may not be
   optimal for all use cases. A weighted aggregation option could be useful.

8. **No `chunk_batches` test for multi-batch scenarios with varying sizes:**
   Existing tests cover `batch_size=1` vs `batch_size=4` but not the case
   where the last batch has fewer items than `batch_size`. The code handles
   this correctly (the `from_fn` loop breaks when chunks are exhausted), but
   an explicit test would be valuable.

---

### Public API Surface

- `Classifier::new(opts)`, `::from_file(path)`, `::from_memory(bytes, level)`
- `::tiny(opts)` — bundled model (feature-gated)
- `classify(samples, top_k)`, `classify_all(samples)`
- `classify_batch(batch, top_k)`, `classify_all_batch(batch)`
- `classify_chunked(samples, top_k, opts)`, `classify_all_chunked(samples, opts)`
- `predict_raw_scores(samples)`, `predict_raw_scores_batch(batch)`
- `predict_raw_scores_batch_flat(batch)`, `predict_raw_scores_batch_into(batch, out)`
- `Options` / `ChunkingOptions` builders
- `EventPrediction` with `index()`, `name()`, `id()`, `confidence()`, `event()`

### Findings

**Good:**
- Comprehensive builder pattern with `with_*` (consuming) and `set_*` (mutable ref)
- Clear separation between raw logits and classified predictions
- `EventPrediction` provides both model-space (index) and ontology-space (event) metadata
- `RankedScore` uses `total_cmp` for NaN-safe ordering of f32 scores
- `BinaryHeap`-based top-k selection is O(n log k) — efficient
- `forbid(unsafe_code)` at crate level
- `#[cfg_attr(not(tarpaulin), inline(always))]` on trivial accessors for coverage tooling

**Observations:**

9. **`&mut self` on all inference methods:** This prevents concurrent inference
   from a single `Classifier` instance. The scratch buffers (`input_scratch`,
   `confidence_scratch`) are the reason. This is a reasonable tradeoff for
   single-threaded throughput but limits concurrency. Consider documenting
   that users should create one `Classifier` per thread if parallelism is
   needed, or add a `Send`/`Sync` note.

10. **No `Display` for `ClassifierError`:** The error derives `thiserror::Error`
    which provides `Display`, but the error messages include inline constants
    like `SAMPLE_RATE_HZ` via string interpolation. This is fine but means
    the error messages contain magic numbers.

11. **`classify_all_batch` returns `Vec<Vec<EventPrediction>>`:** For large
    batch sizes and 527 classes, this creates many small allocations. The
    `predict_raw_scores_batch_flat` + post-processing pattern is more
    efficient but not exposed at the `classify_all_batch` level.

12. **`ChunkingOptions` has no serde round-trip test:** The `test_serde` test
    covers `Options` but not `ChunkingOptions`.

---

### Error Variants

```
ClassifierError::Ort(ort::Error)              — ORT errors
ClassifierError::MissingModelPath             — no path provided
ClassifierError::MissingInputTensor           — model has no input
ClassifierError::MissingOutputTensor          — model has no output
ClassifierError::EmptyInput                   — zero-length audio
ClassifierError::EmptyBatch                   — zero-length batch
ClassifierError::EmptyOutput                  — model returned nothing
ClassifierError::UnexpectedOutputShape        — wrong shape
ClassifierError::UnexpectedClassCount         — wrong number of classes
ClassifierError::MismatchedBatchLength        — clips differ in length
ClassifierError::BatchTooLarge                — allocation overflow
ClassifierError::MissingRatedEventIndex       — unknown class index
ClassifierError::InvalidChunkingOptions       — zero window/hop/batch
```

### Findings

**Good:**
- Comprehensive coverage of failure modes
- `thiserror` for clean derive-based `Display` and `Error` impls
- `#[from]` for ORT error conversion
- `checked_mul` for allocation size overflow protection
- All error variants include relevant context fields
- `ClassifierError` is `Debug` (derived) but not `Clone`/`PartialEq` (correct
  since `ort::Error` likely isn't)

**Observations:**

13. **`BatchTooLarge` is conservative:** The check uses `checked_mul` which
    catches overflow before allocation. However, the actual threshold for
    "too large" is just `usize::MAX`. A configurable max batch size would be
    useful for production deployments to prevent OOM from user input.

14. **`validate_output` has a defensive `.max(1)` (line 972):**
    ```rust
    if raw_scores.len() % expected_batch_size.max(1) == 0 {
    ```
    `expected_batch_size` is always >= 1 at this point (validated upstream),
    so `.max(1)` is redundant but harmless. It does make the code less
    self-documenting.

---

### Design

- `no_std` compatible (edition 2021, MSRV 1.59)
- `phf` perfect hash maps for O(1) case-insensitive lookup
- Two modules: `ontology` (632 entries) and `rated` (527 entries)
- `define_sound_event!` macro generates both types with identical shape
- Codegen via `xtask codegen` from `ontology.json` + `class_labels_indices.csv`

### Findings

**Good:**
- `UncasedStr`-keyed PHF maps for case-insensitive lookups
- Each alias variant generates multiple casings (snake, kebab, camel, etc.)
- Children references filtered to stay within module namespace
- `TryFrom<u64>` for code-based lookup
- `from_index()` for model output class resolution
- `forbid(unsafe_code)` at crate level

**Observations:**

15. **`SipHasher` for code generation:** The `build_record` function uses
    `SipHasher::new().hash(name.as_bytes())`. SipHasher's default
    construction uses a fixed key, making codes deterministic across builds.
    This is correct for the use case but means codes are not portable across
    different hash algorithms.

16. **`Restriction::try_from` is case-sensitive but supports specific variants:**
    ```rust
    "abstract" | "ABSTRACT" | "Abstract" => ...
    "blacklist" | "BLACKLIST" | "BlackList" | "blackList" | "Blacklist" => ...
    ```
    This is a fixed set of known values from the upstream data. Correct but
    slightly inconsistent (not a full case-insensitive match).

---

### Linting

- `#![deny(missing_docs)]` — all public items documented
- `#![forbid(unsafe_code)]` — no unsafe in either crate
- `rust_2018_idioms` and `single_use_lifetimes` warnings enabled
- `unexpected_cfgs` checked for `tarpaulin` and `all_tests`

### Testing

- 41 tests total across feature combinations
- Tests cover: chunking, batching, error paths, serde, builder APIs
- Integration tests with actual ONNX model (bundled-tiny)
- `soundevents-dataset` has 4 integration tests for module structure

### Observations

17. **No doc-tests:** The `cargo test` output shows `running 0 tests` for
    doc-tests. While the crate has good module-level documentation, adding
    runnable examples in doc comments would improve discoverability.

18. **Test helper `pseudo_audio` uses LCG PRNG:** This produces deterministic
    but low-quality pseudo-random audio. Fine for testing but ensures the
    same "audio" is always used, which could mask model sensitivity issues.

---

### Dependencies

| Crate | Version | Purpose |
|-------|---------|---------|
| ort | 2.0.0-rc.12 | ONNX Runtime bindings |
| smol_str | 0.3 | Small string optimization |
| thiserror | 2 | Error derive macro |
| soundevents-dataset | 0.2 | AudioSet ontology data |
| serde | 1 (optional) | Serialization |

### Observations

19. **`ort` is at release candidate:** v2.0.0-rc.12 is not yet stable. The
    crate has a pinned dependency which is good, but API changes between rc
    and stable could require updates.

20. **Optional EP features:** The crate exposes `coreml`, `directml`, `cuda`,
    `rocm`, `tensorrt`, `openvino` features that pass through to `ort`. The
    Cargo.toml comment warns about CoreML producing zeroed logits. This is
    good documentation.

---

### Strengths

- Clean, well-documented public API
- Comprehensive error handling with contextual information
- Efficient memory reuse via scratch buffers
- Strict output validation catches model mismatches
- `forbid(unsafe_code)` in both crates
- Good test coverage including integration tests with real model
- Builder pattern for configuration
- Support for single, batch, and chunked inference

### Items for Improvement

| # | Severity | Description |
|---|----------|-------------|
| 1 | Low | Add doc comment explaining `with_optimization_level` error mapping |
| 2 | Low | Validate/warn if model has multiple I/O tensors |
| 3 | Low | Document first-inference latency (no warmup) |
| 4 | Low | Add sample rate guidance to top-level docs |
| 5 | Low | Document expected amplitude range for input audio |
| 6 | Low | Document tail-chunk behavior for short windows |
| 7 | Low | Consider weighted mean aggregation option for overlapping windows |
| 8 | Low | Add test for chunk_batches with partial last batch |
| 9 | Low | Document `&mut self` concurrency implications |
| 10 | Info | Error messages embed numeric constants inline |
| 11 | Low | Consider flat-output variant of `classify_all_batch` |
| 12 | Low | Add serde round-trip test for `ChunkingOptions` |
| 13 | Low | Consider configurable max batch size limit |
| 14 | Info | Redundant `.max(1)` in `validate_output` |
| 15 | Info | SipHasher codes are fixed-key deterministic |
| 16 | Info | `Restriction::try_from` case matching is partial |
| 17 | Low | Add doc-tests for key public APIs |
| 18 | Info | Test PRNG produces deterministic but limited audio |
| 19 | Low | `ort` dependency is at release candidate |
| 20 | Info | EP feature passthrough with known CoreML issue documented |

### Verdict

**No critical or high-severity issues found.** The codebase is well-engineered
for its purpose. All 41 tests pass on macOS (including bundled ONNX model
inference). The suggestions above are improvements, not blockers.

---

## Test Results

```
Feature set: default          — 25 tests passed
Feature set: bundled-tiny     — 40 tests passed
Feature set: bundled-tiny+serde — 41 tests passed
Feature set: ontology+rated   —  4 tests passed (soundevents-dataset)
```

All 41 tests pass across all feature combinations. No panics, no UB, no
memory safety issues detected.
